### PR TITLE
Metal attachment clears

### PR DESCRIPTION
**Pull request by kvark**
It's great to have that shader-based clear code finally coming handy :)
Known issues: all multi-layered clears fail. Seems to be very similar to #2103. Supposedly, layered rendering in Metal gets totally off the rails when used on an image view, e.g. a texture that has a parent texture with non-zero layer offset. I'll start making a repro case for Apple to address, which will hopefully let us pass all the tests and do it efficiently.

PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [ ] tested examples with the following backends:

### CODE REVIEW

File: src/backend/metal/src/command.rs [hunk 1]
```diff
@@ -1435,15 +1437,179 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
         T: IntoIterator,
         T::Item: Borrow<com::AttachmentClear>,
         U: IntoIterator,
         U::Item: Borrow<pso::ClearRect>,
     {
-        unimplemented!()
+        // gather vertices/polygons
+        let de = self.state.framebuffer_inner.extent;
+        let mut vertices = Vec::new();
+        println!("xx");
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: Should remove this

File: src/backend/metal/src/internal.rs [hunk 2]
```diff
@@ -116,43 +111,63 @@ impl ServicePipes {
...
+    //TODO: return `Option<metal::DepthStencilState>` instead?
+    pub fn get_depth_stencil(&self, depth: bool, stencil: bool) -> &metal::DepthStencilStateRef {
+        if depth && stencil {
+            &self.ds_write_all_state
+        } else if depth {
+            &self.ds_write_depth_state
+        } else if stencil {
+            &self.ds_write_stencil_state
+        } else {
+            panic!("Can't write no thing!")
```
Comment (grovesNL) [inline src/backend/metal/src/internal.rs]: nit: `nothing`

File: src/backend/metal/src/command.rs [hunk 3]
```diff
@@ -1435,15 +1437,179 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
+                    let raw_value = com::ClearColorRaw::from(value);
+                    commands.push(soft::RenderCommand::BindBufferData {
+                        stage: pso::Stage::Fragment,
+                        index: 0,
+                        bytes: unsafe {
+                            slice::from_raw_parts(raw_value.float32.as_ptr() as *const u8, 16)
+                        }.to_owned(),
+                    });
+                    ClearKey {
+                        color: Some((format, index as u8, channel.unwrap())),
```
Comment (grovesNL) [inline src/backend/metal/src/command.rs]: What happens when `index > 1` here? Won't `let ps_name = format!("ps_clear{}_{}", index, s_channel);` fail to find the appropriate shader?

Comment (kvark) [reply src/backend/metal/src/command.rs]: Yep, it feels like this either needs a constant-based RT index selection that comes in Metal 2.0, or mamy many repeated shaders, or some sort of a generator for those (TODO).

**Comment by kvark** [comment]
Thanks for the review! Comments are addressed now, merging.
bors r=grovesNL

File: src/backend/metal/src/command.rs (hunk 3 above)
Comment (grovesNL) [reply src/backend/metal/src/command.rs]: Ok, yeah, maybe a simple generator wouldn't be too hard

**Pull request merged**